fix(core): improve RequestContext type inference for entries/keys/values

abhiaiyer91 opened the pull request:
Fixes the type inference for `RequestContext.entries()`, `keys()`, `values()`, and `forEach()` methods.

Previously, `entries()` returned `[string, union]` which didn't allow type narrowing when iterating. Now it returns a discriminated union of tuples:

```typescript
const ctx = new RequestContext<{ name: string; age: number }>();

for (const [key, value] of ctx.entries()) {
  if (key === 'age') {
    // value is now correctly inferred as `number`
  }
}
```

Fixes #4467

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Improved RequestContext iteration methods so keys(), values(), entries(), and forEach() provide stronger type narrowing and IDE autocompletion for typed contexts.
* **Tests**
  * Added TypeScript type tests to validate precise key/value inference, discriminated entry tuples, nested contexts, and forEach typings.
* **Chores**
  * Added a changelog entry documenting the type-inference improvements.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

abhiaiyer91 committed:
fix(core): improve RequestContext type inference for entries/keys/values

LekoArts commented on packages/core/src/request-context/request-context.test-d.ts:
```diff
@@ -0,0 +1,116 @@
+      context.set('age', 'invalid');
+
+      // @ts-expect-error - unknown key should not be allowed
+      context.set('unknownKey', 'value');
```
That's not using vitest's type testing. Try this (untested) code instead:

```ts
it('should accept correct value types in set()', () => {
  const context = new RequestContext<MyContext>();

  // Verify correct types work
  context.set('name', 'John');
  context.set('age', 25);
  context.set('isActive', true);

  // Assert the value type for 'age' must be number
  expectTypeOf<string>().not.toMatchTypeOf<Parameters<typeof context.set<'age'>>[1]>();

  // Assert 'unknownKey' is not a valid key
  expectTypeOf<'unknownKey'>().not.toMatchTypeOf<keyof MyContext>();
});
```

abhiaiyer91 committed:
fixes

roaminro committed:
Merge branch 'main' into violet-peak-69

roaminro committed:
fix: JSON-encode complex query params and fix RequestContext type compatibility

- Update convertQueryValues to JSON-encode arrays and objects for query params
  (server's wrapSchemaForQueryParams expects JSON strings for complex types)
- Fix inngest workflow RequestContext type annotation to resolve TypeScript
  variance issue (RequestContext<Record<string, any>> not assignable to
  RequestContext<unknown>)

roaminro committed:
docs: improve RequestContext changeset and add TypeScript docs section

- Rewrite changeset to be more developer-oriented with clear before/after
- Add new "TypeScript support" section to RequestContext docs showing
  type-safe set/get, typed keys iteration, and type narrowing with entries()
